lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java

```diff
@@ -333,8 +345,41 @@ private JWEObject signEncryptJar(
                         evidenceRequest,
                         context);
 
-        RSAEncrypter rsaEncrypter = new RSAEncrypter(oauthCriConfig.getParsedEncryptionKey());
-        return AuthorizationRequestHelper.createJweObject(rsaEncrypter, signedJWT);
+        RSAKey encKey;
+        if (cachedOAuthCriEncryptionKeys != null
+                && !cachedOAuthCriEncryptionKeys.isEmpty()
+                && cachedOAuthCriEncryptionKeys.get(cri.getId()) != null
+                && !cachedOAuthCriEncryptionKeys.get(cri.getId()).isExpired()) {
+            encKey = cachedOAuthCriEncryptionKeys.get(cri.getId()).key();
+        } else {
+            encKey = oAuthKeyService.getValidEncryptionKey(oauthCriConfig);
+
+            var cacheDuration =
+                    configService.getParameter(ConfigurationVariable.OAUTH_KEY_CACHE_DURATION_MINS)
+                                    == null
+                            ? DEFAULT_OAUTH_CRI_ENC_KEY_CACHE_DURATION_MINUTES
+                            : Integer.parseInt(
+                                    configService.getParameter(
+                                            ConfigurationVariable.OAUTH_KEY_CACHE_DURATION_MINS));
+            cacheEncryptionKey(encKey, cri, cacheDuration);
+        }
+
+        RSAEncrypter rsaEncrypter = new RSAEncrypter(encKey);
+        return AuthorizationRequestHelper.createJweObject(
+                rsaEncrypter, signedJWT, encKey.getKeyID());
+    }
+
+    private static void cacheEncryptionKey(RSAKey key, Cri cri, Integer cacheDuration) {
+        // Initialise cachedOAuthCriEncryptionKeys if still null
+        if (cachedOAuthCriEncryptionKeys == null) {
+            cachedOAuthCriEncryptionKeys = new HashMap<>();
```

Comment:
It's worth noting that this is _not_ thread-safe - if you have multiple threads then the map might be corrupted.

That should be fine in production because each lambda is effectively single-threaded, but could cause issues with the API tests... It might be worth changing this to be a `ConcurrentHashMap` instead.

You could also initialise this variable directly, unless there's a particular reason to do it lazily? It would simplify a bunch of your logic and make it `final`.